Pinot need skip the failed segments to continue reload the rest of the segments

Suggest change:
![image](https://user-images.githubusercontent.com/13634486/138627336-f81c5549-3fa6-4afa-8735-806503694b08.png)

Failed log:
```
2021-10-22 04:04:34.307 ERROR [HelixTask] [HelixTaskExecutor-message_handle_thread] Exception while executing a message. java.lang.RuntimeException: Caught exception while reloading segment:  in table: wme_metrics_customer_report_REALTIME msgId: e881ac8e-f1e3-4533-85c5-4cd60565cc16 type: USER_DEFINE_MSG
java.lang.RuntimeException: Caught exception while reloading segment:  in table: wme_metrics_customer_report_REALTIME
        at org.apache.pinot.server.starter.helix.SegmentMessageHandlerFactory$SegmentReloadMessageHandler.handleMessage(SegmentMessageHandlerFactory.java:161) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:97) [pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:49) [pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_272]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_272]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_272]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_272]
Caused by: java.lang.IllegalStateException: Index directory: /vdb/pinot/data/wme_metrics_customer_report_REALTIME/wme_metrics_customer_report__24__333__20211001T0003Z is not a directory
        at shaded.com.google.common.base.Preconditions.checkState(Preconditions.java:518) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.pinot.server.starter.helix.HelixInstanceDataManager.reloadSegment(HelixInstanceDataManager.java:262) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.pinot.server.starter.helix.HelixInstanceDataManager.reloadAllSegments(HelixInstanceDataManager.java:224) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.pinot.server.starter.helix.SegmentMessageHandlerFactory$SegmentReloadMessageHandler.handleMessage(SegmentMessageHandlerFactory.java:150) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        ... 6 more
2021-10-22 04:04:34.366 ERROR [HelixTask] [HelixTaskExecutor-message_handle_thread] Message execution failed. msgId: e881ac8e-f1e3-4533-85c5-4cd60565cc16, errorMsg: Caught exception while reloading segment:  in table: wme_metrics_customer_report_REALTIME
2021-10-22 04:04:34.375 ERROR [wme_metrics_customer_report_REALTIME-SegmentReloadMessageHandler] [HelixTaskExecutor-message_handle_thread] onError: INTERNAL, ERROR
java.lang.RuntimeException: Caught exception while reloading segment:  in table: wme_metrics_customer_report_REALTIME
        at org.apache.pinot.server.starter.helix.SegmentMessageHandlerFactory$SegmentReloadMessageHandler.handleMessage(SegmentMessageHandlerFactory.java:161) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:97) [pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:49) [pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_272]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_272]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_272]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_272]
Caused by: java.lang.IllegalStateException: Index directory: /vdb/pinot/data/wme_metrics_customer_report_REALTIME/wme_metrics_customer_report__24__333__20211001T0003Z is not a directory
        at shaded.com.google.common.base.Preconditions.checkState(Preconditions.java:518) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.pinot.server.starter.helix.HelixInstanceDataManager.reloadSegment(HelixInstanceDataManager.java:262) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.pinot.server.starter.helix.HelixInstanceDataManager.reloadAllSegments(HelixInstanceDataManager.java:224) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        at org.apache.pinot.server.starter.helix.SegmentMessageHandlerFactory$SegmentReloadMessageHandler.handleMessage(SegmentMessageHandlerFactory.java:150) ~[pinot-all-0.8.0-jar-with-dependencies.jar:0.8.0-031903bac9a6b6fa3b906d6093925d5fdc2e5502]
        ... 6 more
```

--- a/pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java
+++ b/pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java
@@ -168,9 +168,9 @@ public class HelixInstanceDataManager implements InstanceDataManager {
     TableDataManagerConfig tableDataManagerConfig =
         TableDataManagerConfig.getDefaultHelixTableDataManagerConfig(_instanceDataManagerConfig, tableNameWithType);
     tableDataManagerConfig.overrideConfigs(tableConfig, _authToken);
-    TableDataManager tableDataManager = TableDataManagerProvider
-        .getTableDataManager(tableDataManagerConfig, _instanceId, _propertyStore, _serverMetrics, _helixManager,
-            _errorCache);
+    TableDataManager tableDataManager =
+        TableDataManagerProvider.getTableDataManager(tableDataManagerConfig, _instanceId, _propertyStore,
+            _serverMetrics, _helixManager, _errorCache);
     tableDataManager.start();
     LOGGER.info("Created table data manager for table: {}", tableNameWithType);
     return tableDataManager;
@@ -212,16 +212,31 @@ public class HelixInstanceDataManager implements InstanceDataManager {
   }
 
   @Override
-  public void reloadAllSegments(String tableNameWithType, boolean forceDownload)
-      throws Exception {
+  public void reloadAllSegments(String tableNameWithType, boolean forceDownload) {
     LOGGER.info("Reloading all segments in table: {}", tableNameWithType);
     TableConfig tableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, tableNameWithType);
     Preconditions.checkNotNull(tableConfig);
 
     Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, tableNameWithType);
 
-    for (SegmentMetadata segmentMetadata : getAllSegmentsMetadata(tableNameWithType)) {
-      reloadSegment(tableNameWithType, segmentMetadata, tableConfig, schema, forceDownload);
+    List<String> failedSegments = new ArrayList<>();
+    Exception sampleException = null;
+    List<SegmentMetadata> segmentsMetadata = getAllSegmentsMetadata(tableNameWithType);
+    for (SegmentMetadata segmentMetadata : segmentsMetadata) {
+      try {
+        reloadSegment(tableNameWithType, segmentMetadata, tableConfig, schema, forceDownload);
+      } catch (Exception e) {
+        String segmentName = segmentMetadata.getName();
+        LOGGER.error("Caught exception while reloading segment: {} in table: {}", segmentName, tableNameWithType, e);
+        failedSegments.add(segmentName);
+        sampleException = e;
+      }
+    }
+
+    if (sampleException != null) {
+      throw new RuntimeException(
+          String.format("Failed to reload %d/%d segments: %s in table: %s", failedSegments.size(),
+              segmentsMetadata.size(), failedSegments, tableNameWithType), sampleException);
     }
 
     LOGGER.info("Reloaded all segments in table: {}", tableNameWithType);
@@ -300,8 +315,8 @@ public class HelixInstanceDataManager implements InstanceDataManager {
       SegmentMetadata localMetadata = getSegmentMetadata(tableNameWithType, segmentName);
 
       _tableDataManagerMap.computeIfAbsent(tableNameWithType, k -> createTableDataManager(k, tableConfig))
-          .addOrReplaceSegment(segmentName, new IndexLoadingConfig(_instanceDataManagerConfig, tableConfig),
-              zkMetadata, localMetadata);
+          .addOrReplaceSegment(segmentName, new IndexLoadingConfig(_instanceDataManagerConfig, tableConfig), zkMetadata,
+              localMetadata);
       LOGGER.info("Added or replaced segment: {} of table: {}", segmentName, tableNameWithType);
     } finally {
       segmentLock.unlock();

